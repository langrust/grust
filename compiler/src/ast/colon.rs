use syn::{parse::Parse, Token};

pub struct Colon<U: Parse, V: Parse> {
    pub left: U,
    pub colon: Token![:],
    pub right: V,
}
impl<U: Parse, V: Parse> Parse for Colon<U, V> {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        Ok(Self {
            left: input.parse()?,
            colon: input.parse()?,
            right: input.parse()?,
        })
    }
}
