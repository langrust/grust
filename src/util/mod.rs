/// Location handler module.
pub mod location;

/// Files handler module.
pub mod files;

/// Type system module.
pub mod type_system;

/// Constant module.
pub mod constants;
