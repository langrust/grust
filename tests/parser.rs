#[cfg(test)]
mod langrust_ast_constructs {
    use codespan_reporting::files::Files;
    use grustine::ast::{
        component::Component, expression::Expression, file::File, function::Function, node::Node,
        stream_expression::StreamExpression, user_defined_type::UserDefinedType,
    };
    use grustine::langrust;
    use grustine::util::{
        constant::Constant,
        files,
        location::Location,
        operator::{BinaryOperator, OtherOperator, UnaryOperator},
        type_system::Type,
    };

    #[test]
    fn file_parser() {
        let mut files = files::Files::new();

        let module_test_id = files
            .add("module_test.gr", "function node enum node function node")
            .unwrap();
        let program_test_id = files
            .add(
                "program_test.gr",
                "node component array node function struct function",
            )
            .unwrap();

        let file = langrust::fileParser::new()
            .parse(module_test_id, &files.source(module_test_id).unwrap())
            .unwrap();
        assert_eq!(
            file,
            File::Module {
                user_defined_types: vec![UserDefinedType::Enumeration {
                    location: Location::default()
                }],
                functions: vec![
                    Function {
                        location: Location::default()
                    },
                    Function {
                        location: Location::default()
                    }
                ],
                nodes: vec![
                    Node {
                        location: Location::default()
                    },
                    Node {
                        location: Location::default()
                    },
                    Node {
                        location: Location::default()
                    }
                ],
                location: Location::default()
            },
        );

        let file = langrust::fileParser::new()
            .parse(program_test_id, &files.source(program_test_id).unwrap())
            .unwrap();
        assert_eq!(
            file,
            File::Program {
                user_defined_types: vec![
                    UserDefinedType::Array {
                        location: Location::default()
                    },
                    UserDefinedType::Structure {
                        location: Location::default()
                    }
                ],
                functions: vec![
                    Function {
                        location: Location::default()
                    },
                    Function {
                        location: Location::default()
                    }
                ],
                nodes: vec![
                    Node {
                        location: Location::default()
                    },
                    Node {
                        location: Location::default()
                    }
                ],
                component: Component {
                    location: Location::default()
                },
                location: Location::default()
            },
        );
    }

    #[test]
    fn types() {
        let mut files = files::Files::new();
        let file_id1 = files.add("int_test.gr", "int").unwrap();
        let file_id2 = files.add("float_test.gr", "float").unwrap();
        let file_id3 = files.add("bool_test.gr", "bool").unwrap();
        let file_id4 = files.add("string_test.gr", "string").unwrap();
        let file_id5 = files.add("unit_test.gr", "unit").unwrap();
        let file_id6 = files.add("array_test.gr", "[int; 3]").unwrap();
        let file_id7 = files.add("option_test.gr", "int?").unwrap();
        let file_id8 = files.add("undefined_type_test.gr", "Color").unwrap();

        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id1, &files.source(file_id1).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Integer);
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id2, &files.source(file_id2).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Float);
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id3, &files.source(file_id3).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Boolean);
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id4, &files.source(file_id4).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::String);
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id5, &files.source(file_id5).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Unit);
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id6, &files.source(file_id6).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Array(Box::new(Type::Integer), 3));
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id7, &files.source(file_id7).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::Option(Box::new(Type::Integer)));
        let basic_type = langrust::basicTypeParser::new()
            .parse(file_id8, &files.source(file_id8).unwrap())
            .unwrap();
        assert_eq!(basic_type, Type::NotDefinedYet(String::from("Color")));
    }

    #[test]
    fn stream_expression() {
        let mut files = files::Files::new();
        let file_id1 = files.add("constant_test.gr", "Color.Yellow").unwrap();
        let file_id2 = files.add("signal_call_test.gr", "x").unwrap();
        let file_id3 = files.add("brackets_test.gr", "(3)").unwrap();
        let file_id4 = files.add("unary_test.gr", "-3").unwrap();
        let file_id5 = files.add("binary_test.gr", "4*5-3").unwrap();
        let file_id6 = files
            .add("map_application_test.gr", "(x*y).map(sqrt)")
            .unwrap();
        let file_id7 = files
            .add("print_test.gr", "print(\"Hello world\")")
            .unwrap();
        let file_id8 = files
            .add(
                "node_application_test.gr",
                "my_node(my_input1, my_input2).my_signal",
            )
            .unwrap();
        let file_id9 = files.add("fby_test.gr", "0 fby x + 1").unwrap();
        let file_id10 = files
            .add("ifthenelse_test.gr", "if b then x else y")
            .unwrap();
        let file_id11 = files
            .add("struct_test.gr", "Point { x: 3, y: 0, }")
            .unwrap();
        let file_id12 = files.add("array_test.gr", "[1, 2, 3]").unwrap();
        let file_id13 = files.add("unified_array_test.gr", "[0.01; 3]").unwrap();
        let file_id14 = files
            .add("when_id_test.gr", "when a = x then a else 0")
            .unwrap();
        let file_id15 = files.add("when_test.gr", "when a then a else 0").unwrap();

        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id1, &files.source(file_id1).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::Constant {
                constant: Constant::Enumeration(String::from("Color"), String::from("Yellow")),
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id2, &files.source(file_id2).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::SignalCall {
                id: String::from("x"),
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id3, &files.source(file_id3).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: UnaryOperator::Brackets.to_string(),
                    location: Location::default()
                },
                inputs: vec![StreamExpression::Constant {
                    constant: Constant::Integer(3),
                    location: Location::default()
                },],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id4, &files.source(file_id4).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: UnaryOperator::Neg.to_string(),
                    location: Location::default()
                },
                inputs: vec![StreamExpression::Constant {
                    constant: Constant::Integer(3),
                    location: Location::default()
                },],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id5, &files.source(file_id5).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: BinaryOperator::Sub.to_string(),
                    location: Location::default()
                },
                inputs: vec![
                    StreamExpression::MapApplication {
                        expression: Expression::Call {
                            id: BinaryOperator::Mul.to_string(),
                            location: Location::default()
                        },
                        inputs: vec![
                            StreamExpression::Constant {
                                constant: Constant::Integer(4),
                                location: Location::default()
                            },
                            StreamExpression::Constant {
                                constant: Constant::Integer(5),
                                location: Location::default()
                            },
                        ],
                        location: Location::default()
                    },
                    StreamExpression::Constant {
                        constant: Constant::Integer(3),
                        location: Location::default()
                    },
                ],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id6, &files.source(file_id6).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: String::from("sqrt"),
                    location: Location::default()
                },
                inputs: vec![StreamExpression::MapApplication {
                    expression: Expression::Call {
                        id: BinaryOperator::Mul.to_string(),
                        location: Location::default()
                    },
                    inputs: vec![
                        StreamExpression::SignalCall {
                            id: String::from("x"),
                            location: Location::default()
                        },
                        StreamExpression::SignalCall {
                            id: String::from("y"),
                            location: Location::default()
                        },
                    ],
                    location: Location::default()
                },],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id7, &files.source(file_id7).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: OtherOperator::Print.to_string(),
                    location: Location::default()
                },
                inputs: vec![StreamExpression::Constant {
                    constant: Constant::String(String::from("Hello world")),
                    location: Location::default()
                }],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id8, &files.source(file_id8).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::NodeApplication {
                node: String::from("my_node"),
                inputs: vec![
                    StreamExpression::SignalCall {
                        id: String::from("my_input1"),
                        location: Location::default()
                    },
                    StreamExpression::SignalCall {
                        id: String::from("my_input2"),
                        location: Location::default()
                    }
                ],
                signal: String::from("my_signal"),
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id9, &files.source(file_id9).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::FollowedBy {
                constant: Constant::Integer(0),
                expression: Box::new(StreamExpression::MapApplication {
                    expression: Expression::Call {
                        id: BinaryOperator::Add.to_string(),
                        location: Location::default()
                    },
                    inputs: vec![
                        StreamExpression::SignalCall {
                            id: String::from("x"),
                            location: Location::default()
                        },
                        StreamExpression::Constant {
                            constant: Constant::Integer(1),
                            location: Location::default()
                        },
                    ],
                    location: Location::default()
                }),
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id10, &files.source(file_id10).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::MapApplication {
                expression: Expression::Call {
                    id: OtherOperator::IfThenElse.to_string(),
                    location: Location::default()
                },
                inputs: vec![
                    StreamExpression::SignalCall {
                        id: String::from("b"),
                        location: Location::default()
                    },
                    StreamExpression::SignalCall {
                        id: String::from("x"),
                        location: Location::default()
                    },
                    StreamExpression::SignalCall {
                        id: String::from("y"),
                        location: Location::default()
                    },
                ],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id11, &files.source(file_id11).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::Structure {
                name: String::from("Point"),
                fields: vec![
                    (
                        String::from("x"),
                        StreamExpression::Constant {
                            constant: Constant::Integer(3),
                            location: Location::default()
                        }
                    ),
                    (
                        String::from("y"),
                        StreamExpression::Constant {
                            constant: Constant::Integer(0),
                            location: Location::default()
                        }
                    )
                ],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id12, &files.source(file_id12).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::Array {
                elements: vec![
                    StreamExpression::Constant {
                        constant: Constant::Integer(1),
                        location: Location::default()
                    },
                    StreamExpression::Constant {
                        constant: Constant::Integer(2),
                        location: Location::default()
                    },
                    StreamExpression::Constant {
                        constant: Constant::Integer(3),
                        location: Location::default()
                    }
                ],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id13, &files.source(file_id13).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::Array {
                elements: vec![
                    StreamExpression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    },
                    StreamExpression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    },
                    StreamExpression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    }
                ],
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id14, &files.source(file_id14).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::When {
                id: String::from("a"),
                option: Box::new(StreamExpression::SignalCall {
                    id: String::from("x"),
                    location: Location::default()
                }),
                present: Box::new(StreamExpression::SignalCall {
                    id: String::from("a"),
                    location: Location::default()
                }),
                default: Box::new(StreamExpression::Constant {
                    constant: Constant::Integer(0),
                    location: Location::default()
                }),
                location: Location::default()
            },
            stream_expression
        );
        let stream_expression = langrust::streamExpressionParser::new()
            .parse(file_id15, &files.source(file_id15).unwrap())
            .unwrap();
        assert_eq!(
            StreamExpression::When {
                id: String::from("a"),
                option: Box::new(StreamExpression::SignalCall {
                    id: String::from("a"),
                    location: Location::default()
                }),
                present: Box::new(StreamExpression::SignalCall {
                    id: String::from("a"),
                    location: Location::default()
                }),
                default: Box::new(StreamExpression::Constant {
                    constant: Constant::Integer(0),
                    location: Location::default()
                }),
                location: Location::default()
            },
            stream_expression
        );
    }

    #[test]
    fn expression() {
        let mut files = files::Files::new();
        let file_id1 = files.add("constant_test.gr", "Color.Yellow").unwrap();
        let file_id2 = files.add("element_call_test.gr", "x").unwrap();
        let file_id3 = files.add("brackets_test.gr", "(3)").unwrap();
        let file_id4 = files.add("unary_test.gr", "-3").unwrap();
        let file_id5 = files.add("binary_test.gr", "4*5-3").unwrap();
        let file_id6 = files
            .add("function_application_test.gr", "sqrt(4*5-3)")
            .unwrap();
        let file_id7 = files
            .add("print_test.gr", "print(\"Hello world\")")
            .unwrap();
        let file_id8 = files.add("abstraction_test.gr", "|x, y| x + y").unwrap();
        let file_id9 = files
            .add("typed_abstraction_test.gr", "|x: int, y: int| x + y")
            .unwrap();
        let file_id10 = files
            .add("ifthenelse_test.gr", "if b then x else y")
            .unwrap();
        let file_id11 = files
            .add("struct_test.gr", "Point { x: 3, y: 0, }")
            .unwrap();
        let file_id12 = files.add("array_test.gr", "[1, 2, 3]").unwrap();
        let file_id13 = files.add("unified_array_test.gr", "[0.01; 3]").unwrap();
        let file_id14 = files
            .add("when_id_test.gr", "when a = x then a else 0")
            .unwrap();
        let file_id15 = files.add("when_test.gr", "when a then a else 0").unwrap();

        let expression = langrust::expressionParser::new()
            .parse(file_id1, &files.source(file_id1).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Constant {
                constant: Constant::Enumeration(String::from("Color"), String::from("Yellow")),
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id2, &files.source(file_id2).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Call {
                id: String::from("x"),
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id3, &files.source(file_id3).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: UnaryOperator::Brackets.to_string(),
                    location: Location::default()
                }),
                inputs: vec![Expression::Constant {
                    constant: Constant::Integer(3),
                    location: Location::default()
                },],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id4, &files.source(file_id4).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: UnaryOperator::Neg.to_string(),
                    location: Location::default()
                }),
                inputs: vec![Expression::Constant {
                    constant: Constant::Integer(3),
                    location: Location::default()
                },],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id5, &files.source(file_id5).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: BinaryOperator::Sub.to_string(),
                    location: Location::default()
                }),
                inputs: vec![
                    Expression::Application {
                        expression: Box::new(Expression::Call {
                            id: BinaryOperator::Mul.to_string(),
                            location: Location::default()
                        }),
                        inputs: vec![
                            Expression::Constant {
                                constant: Constant::Integer(4),
                                location: Location::default()
                            },
                            Expression::Constant {
                                constant: Constant::Integer(5),
                                location: Location::default()
                            },
                        ],
                        location: Location::default()
                    },
                    Expression::Constant {
                        constant: Constant::Integer(3),
                        location: Location::default()
                    },
                ],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id6, &files.source(file_id6).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: String::from("sqrt"),
                    location: Location::default()
                }),
                inputs: vec![Expression::Application {
                    expression: Box::new(Expression::Call {
                        id: BinaryOperator::Sub.to_string(),
                        location: Location::default()
                    }),
                    inputs: vec![
                        Expression::Application {
                            expression: Box::new(Expression::Call {
                                id: BinaryOperator::Mul.to_string(),
                                location: Location::default()
                            }),
                            inputs: vec![
                                Expression::Constant {
                                    constant: Constant::Integer(4),
                                    location: Location::default()
                                },
                                Expression::Constant {
                                    constant: Constant::Integer(5),
                                    location: Location::default()
                                },
                            ],
                            location: Location::default()
                        },
                        Expression::Constant {
                            constant: Constant::Integer(3),
                            location: Location::default()
                        },
                    ],
                    location: Location::default()
                }],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id7, &files.source(file_id7).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: OtherOperator::Print.to_string(),
                    location: Location::default()
                }),
                inputs: vec![Expression::Constant {
                    constant: Constant::String(String::from("Hello world")),
                    location: Location::default()
                }],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id8, &files.source(file_id8).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Abstraction {
                inputs: vec![String::from("x"), String::from("y")],
                expression: Box::new(Expression::Application {
                    expression: Box::new(Expression::Call {
                        id: BinaryOperator::Add.to_string(),
                        location: Location::default()
                    }),
                    inputs: vec![
                        Expression::Call {
                            id: String::from("x"),
                            location: Location::default()
                        },
                        Expression::Call {
                            id: String::from("y"),
                            location: Location::default()
                        },
                    ],
                    location: Location::default()
                }),
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id9, &files.source(file_id9).unwrap())
            .unwrap();
        assert_eq!(
            Expression::TypedAbstraction {
                inputs: vec![
                    (String::from("x"), Type::Integer),
                    (String::from("y"), Type::Integer)
                ],
                expression: Box::new(Expression::Application {
                    expression: Box::new(Expression::Call {
                        id: BinaryOperator::Add.to_string(),
                        location: Location::default()
                    }),
                    inputs: vec![
                        Expression::Call {
                            id: String::from("x"),
                            location: Location::default()
                        },
                        Expression::Call {
                            id: String::from("y"),
                            location: Location::default()
                        },
                    ],
                    location: Location::default()
                }),
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id10, &files.source(file_id10).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Application {
                expression: Box::new(Expression::Call {
                    id: OtherOperator::IfThenElse.to_string(),
                    location: Location::default()
                }),
                inputs: vec![
                    Expression::Call {
                        id: String::from("b"),
                        location: Location::default()
                    },
                    Expression::Call {
                        id: String::from("x"),
                        location: Location::default()
                    },
                    Expression::Call {
                        id: String::from("y"),
                        location: Location::default()
                    },
                ],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id11, &files.source(file_id11).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Structure {
                name: String::from("Point"),
                fields: vec![
                    (
                        String::from("x"),
                        Expression::Constant {
                            constant: Constant::Integer(3),
                            location: Location::default()
                        }
                    ),
                    (
                        String::from("y"),
                        Expression::Constant {
                            constant: Constant::Integer(0),
                            location: Location::default()
                        }
                    )
                ],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id12, &files.source(file_id12).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Array {
                elements: vec![
                    Expression::Constant {
                        constant: Constant::Integer(1),
                        location: Location::default()
                    },
                    Expression::Constant {
                        constant: Constant::Integer(2),
                        location: Location::default()
                    },
                    Expression::Constant {
                        constant: Constant::Integer(3),
                        location: Location::default()
                    }
                ],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id13, &files.source(file_id13).unwrap())
            .unwrap();
        assert_eq!(
            Expression::Array {
                elements: vec![
                    Expression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    },
                    Expression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    },
                    Expression::Constant {
                        constant: Constant::Float(0.01),
                        location: Location::default()
                    }
                ],
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id14, &files.source(file_id14).unwrap())
            .unwrap();
        assert_eq!(
            Expression::When {
                id: String::from("a"),
                option: Box::new(Expression::Call {
                    id: String::from("x"),
                    location: Location::default()
                }),
                present: Box::new(Expression::Call {
                    id: String::from("a"),
                    location: Location::default()
                }),
                default: Box::new(Expression::Constant {
                    constant: Constant::Integer(0),
                    location: Location::default()
                }),
                location: Location::default()
            },
            expression
        );
        let expression = langrust::expressionParser::new()
            .parse(file_id15, &files.source(file_id15).unwrap())
            .unwrap();
        assert_eq!(
            Expression::When {
                id: String::from("a"),
                option: Box::new(Expression::Call {
                    id: String::from("a"),
                    location: Location::default()
                }),
                present: Box::new(Expression::Call {
                    id: String::from("a"),
                    location: Location::default()
                }),
                default: Box::new(Expression::Constant {
                    constant: Constant::Integer(0),
                    location: Location::default()
                }),
                location: Location::default()
            },
            expression
        );
    }

    #[test]
    fn constant() {
        let mut files = files::Files::new();
        let file_id1 = files.add("unit_test.gr", "()").unwrap();
        let file_id2 = files.add("bool_test.gr", "true").unwrap();
        let file_id3 = files.add("int_test.gr", "3").unwrap();
        let file_id4 = files.add("float_test.gr", "3.540").unwrap();
        let file_id5 = files.add("string_test.gr", "\"Hello world\"").unwrap();
        let file_id6 = files.add("enum_test.gr", "Color.Yellow").unwrap();

        let constant = langrust::constantParser::new()
            .parse(file_id1, &files.source(file_id1).unwrap())
            .unwrap();
        assert_eq!(Constant::Unit, constant);
        let constant = langrust::constantParser::new()
            .parse(file_id2, &files.source(file_id2).unwrap())
            .unwrap();
        assert_eq!(Constant::Boolean(true), constant);
        let constant = langrust::constantParser::new()
            .parse(file_id3, &files.source(file_id3).unwrap())
            .unwrap();
        assert_eq!(Constant::Integer(3), constant);
        let constant = langrust::constantParser::new()
            .parse(file_id4, &files.source(file_id4).unwrap())
            .unwrap();
        assert_eq!(Constant::Float(3.540), constant);
        let constant = langrust::constantParser::new()
            .parse(file_id5, &files.source(file_id5).unwrap())
            .unwrap();
        assert_eq!(Constant::String(String::from("Hello world")), constant);
        let constant = langrust::constantParser::new()
            .parse(file_id6, &files.source(file_id6).unwrap())
            .unwrap();
        assert_eq!(
            Constant::Enumeration(String::from("Color"), String::from("Yellow")),
            constant
        );
    }
}
