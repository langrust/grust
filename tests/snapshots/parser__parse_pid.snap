---
source: tests/parser.rs
expression: file
---
typedefs:
  - Structure:
      id: GainPID
      fields:
        - - k_p
          - Float
        - - k_i
          - Float
        - - k_d
          - Float
      location: ~
functions:
  - id: min
    inputs:
      - - a
        - Float
      - - b
        - Float
    statements:
      - id: test
        element_type: Boolean
        expression:
          Application:
            function_expression:
              Call:
                id: " > "
                typing: ~
                location: ~
            inputs:
              - Call:
                  id: a
                  typing: ~
                  location: ~
              - Call:
                  id: b
                  typing: ~
                  location: ~
            typing: ~
            location: ~
        location: ~
      - id: result
        element_type: Float
        expression:
          Application:
            function_expression:
              Call:
                id: if_then_else
                typing: ~
                location: ~
            inputs:
              - Call:
                  id: test
                  typing: ~
                  location: ~
              - Call:
                  id: b
                  typing: ~
                  location: ~
              - Call:
                  id: a
                  typing: ~
                  location: ~
            typing: ~
            location: ~
        location: ~
    returned:
      - Float
      - Call:
          id: result
          typing: ~
          location: ~
    location: ~
nodes:
  - id: integrate
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    equations:
      - - prev_i
        - scope: Local
          id: prev_i
          signal_type: Float
          expression:
            FollowedBy:
              constant:
                Float: 0
              expression:
                SignalCall:
                  id: i
                  typing: ~
                  location: ~
              typing: ~
              location: ~
          location: ~
      - - i
        - scope: Output
          id: i
          signal_type: Float
          expression:
            MapApplication:
              function_expression:
                Call:
                  id: min
                  typing: ~
                  location: ~
              inputs:
                - MapApplication:
                    function_expression:
                      Call:
                        id: " + "
                        typing: ~
                        location: ~
                    inputs:
                      - SignalCall:
                          id: prev_i
                          typing: ~
                          location: ~
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " * "
                              typing: ~
                              location: ~
                          inputs:
                            - SignalCall:
                                id: x
                                typing: ~
                                location: ~
                            - SignalCall:
                                id: dt
                                typing: ~
                                location: ~
                          typing: ~
                          location: ~
                    typing: ~
                    location: ~
                - SignalCall:
                    id: MAX_INTEGRALE
                    typing: ~
                    location: ~
              typing: ~
              location: ~
          location: ~
      - - MAX_INTEGRALE
        - scope: Local
          id: MAX_INTEGRALE
          signal_type: Float
          expression:
            Constant:
              constant:
                Float: 1000
              typing: ~
              location: ~
          location: ~
    location: ~
  - id: derive
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    equations:
      - - d
        - scope: Output
          id: d
          signal_type: Float
          expression:
            MapApplication:
              function_expression:
                Call:
                  id: " / "
                  typing: ~
                  location: ~
              inputs:
                - MapApplication:
                    function_expression:
                      Call:
                        id: (_)
                        typing: ~
                        location: ~
                    inputs:
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " - "
                              typing: ~
                              location: ~
                          inputs:
                            - SignalCall:
                                id: x
                                typing: ~
                                location: ~
                            - SignalCall:
                                id: prev_x
                                typing: ~
                                location: ~
                          typing: ~
                          location: ~
                    typing: ~
                    location: ~
                - SignalCall:
                    id: dt
                    typing: ~
                    location: ~
              typing: ~
              location: ~
          location: ~
      - - prev_x
        - scope: Local
          id: prev_x
          signal_type: Float
          expression:
            FollowedBy:
              constant:
                Float: 0
              expression:
                SignalCall:
                  id: x
                  typing: ~
                  location: ~
              typing: ~
              location: ~
          location: ~
    location: ~
component:
  id: pid
  is_component: true
  inputs:
    - - v_c
      - Float
    - - v
      - Float
    - - dt
      - Float
  equations:
    - - gain
      - scope: Local
        id: gain
        signal_type:
          NotDefinedYet: GainPID
        expression:
          Structure:
            name: GainPID
            fields:
              - - k_p
                - Constant:
                    constant:
                      Float: 0.2
                    typing: ~
                    location: ~
              - - k_i
                - Constant:
                    constant:
                      Float: 1.5
                    typing: ~
                    location: ~
              - - k_d
                - Constant:
                    constant:
                      Float: 6
                    typing: ~
                    location: ~
            typing: ~
            location: ~
        location: ~
    - - u
      - scope: Output
        id: u
        signal_type: Float
        expression:
          Match:
            expression:
              SignalCall:
                id: gain
                typing: ~
                location: ~
            arms:
              - - Structure:
                    name: GainPID
                    fields:
                      - - k_p
                        - Identifier:
                            name: k_p
                            location: ~
                      - - k_i
                        - Identifier:
                            name: k_i
                            location: ~
                      - - k_d
                        - Identifier:
                            name: k_d
                            location: ~
                    location: ~
                - ~
                - MapApplication:
                    function_expression:
                      Call:
                        id: " + "
                        typing: ~
                        location: ~
                    inputs:
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " + "
                              typing: ~
                              location: ~
                          inputs:
                            - MapApplication:
                                function_expression:
                                  Call:
                                    id: " * "
                                    typing: ~
                                    location: ~
                                inputs:
                                  - SignalCall:
                                      id: k_p
                                      typing: ~
                                      location: ~
                                  - SignalCall:
                                      id: e
                                      typing: ~
                                      location: ~
                                typing: ~
                                location: ~
                            - MapApplication:
                                function_expression:
                                  Call:
                                    id: " * "
                                    typing: ~
                                    location: ~
                                inputs:
                                  - SignalCall:
                                      id: k_i
                                      typing: ~
                                      location: ~
                                  - SignalCall:
                                      id: e_i
                                      typing: ~
                                      location: ~
                                typing: ~
                                location: ~
                          typing: ~
                          location: ~
                      - MapApplication:
                          function_expression:
                            Call:
                              id: " * "
                              typing: ~
                              location: ~
                          inputs:
                            - SignalCall:
                                id: k_d
                                typing: ~
                                location: ~
                            - SignalCall:
                                id: e_d
                                typing: ~
                                location: ~
                          typing: ~
                          location: ~
                    typing: ~
                    location: ~
            typing: ~
            location: ~
        location: ~
    - - e
      - scope: Local
        id: e
        signal_type: Float
        expression:
          MapApplication:
            function_expression:
              Call:
                id: " - "
                typing: ~
                location: ~
            inputs:
              - SignalCall:
                  id: v_c
                  typing: ~
                  location: ~
              - SignalCall:
                  id: v
                  typing: ~
                  location: ~
            typing: ~
            location: ~
        location: ~
    - - e_i
      - scope: Local
        id: e_i
        signal_type: Float
        expression:
          NodeApplication:
            node: integrate
            inputs:
              - SignalCall:
                  id: e
                  typing: ~
                  location: ~
              - SignalCall:
                  id: dt
                  typing: ~
                  location: ~
            signal: i
            typing: ~
            location: ~
        location: ~
    - - e_d
      - scope: Local
        id: e_d
        signal_type: Float
        expression:
          NodeApplication:
            node: derive
            inputs:
              - SignalCall:
                  id: e
                  typing: ~
                  location: ~
              - SignalCall:
                  id: dt
                  typing: ~
                  location: ~
            signal: d
            typing: ~
            location: ~
        location: ~
  location: ~
location: ~

